packages/api/src/services/emp-stats.ts

```diff
@@ -34,51 +33,76 @@ export default (config: Config, appState: Dependencies) => {
-    const update = {
-      address: stat.address,
-      tvl: stat.tvl,
-      tvm: stat.tvm,
-      timestamp: stat.timestamp,
-    };
-    return createStatHistory(address, update);
+  async function updateHistory(empAddress: string) {
+    const stat = await getLatestStats(empAddress);
+    assert(uma.utils.exists(stat.timestamp), "stats require timestamp for emp: " + empAddress);
+    if (await hasStatsHistory(empAddress, stat.timestamp)) return stat;
+    // remove id from the latest stats, since this gets overriden with history stats by timestamp
+    const { id, ...historyStat } = stat;
+    return createStatHistory(empAddress, historyStat as EmpHistoryStat);
   }
-
   async function updateHistories(addresses: string[]) {
     return Promise.allSettled(addresses.map(updateHistory));
   }
 
-  async function updateLatestTvl(address: string) {
-    const emp = await queries.getAnyEmp(address);
-    // the full state has collateral decimals, pulled from erc20 state
-    const fullState = await queries.getFullEmpState(emp);
+  // synthetic prices return price in wei, due to the way the bot price feeds work
+  async function getSyntheticPriceFromTable(empAddress: string) {
+    // PriceSample is type [ timestamp:number, price:string]
+    const priceSample: PriceSample = synthPrices.latest[empAddress];
+    assert(uma.utils.exists(priceSample), "No latest synthetic price found for emp: " + empAddress);
 
-    assert(uma.utils.exists(fullState.collateralCurrency), "Emp requires collateralCurrency: " + address);
+    const price = priceSample[1];
+    assert(uma.utils.exists(price), "Invalid latest sythetic price found on emp: " + empAddress);
+    return price;
+  }
 
+  async function getPriceFromTable(empAddress: string, currencyAddress: string) {
     // PriceSample is type [ timestamp:number, price:string]
-    const priceSample: PriceSample = prices[currency].latest[fullState.collateralCurrency];
-    assert(uma.utils.exists(priceSample), "No latest price found for emp: " + address);
-    const [timestamp, price] = priceSample;
-    assert(uma.utils.exists(price), "Invalid latest price found on emp: " + address);
+    const priceSample: PriceSample = prices[currency].latest[currencyAddress];
+    assert(uma.utils.exists(priceSample), "No latest price found for emp: " + empAddress);
+    const [, price] = priceSample;
+    assert(uma.utils.exists(price), "Invalid latest price found on emp: " + empAddress);
+    return price;
+  }
+
+  async function getFullEmpState(empAddress: string) {
+    const emp = await queries.getAnyEmp(empAddress);
+    // the full state has collateral decimals, pulled from erc20 state
+    return queries.getFullEmpState(emp);
+  }
+  async function updateStats(empAddress: string) {
+    const update: empStats.Data = { address: empAddress };
+
+    const emp = await getFullEmpState(empAddress);
+    update.timestamp = nowS();
+
+    if (uma.utils.exists(emp.collateralCurrency)) {
```

Comment:
When would this be false? Same question for tokenCurrency below? Unless I'm missing something, that would signal that something's broken?

Reply:
theres 2 ways i can think of this is undefined. 

1. The the state for emps is not consistent, its possible to have state come  in over time, so if this happens to run before an emp is fully fetched, it could be undefined. This is not super likely because we wait for emp state to fetch before this is ever run, but you never know. 
2. More likely is that currently the system cant differentiate between non emp contracts deployed through the registry. It assumes all registry contract events are EMPs, but actually some of these contracts are like governance contracts, or even early test emp contracts which dont have this collateral address.  So this is iterating over some contract addresses that are not actually EMPs, which cause problems. 

Dont' really have a good solution for #2 without someway to whitelist/blacklist addresses or filter them based on some kind of criteria

Reply:
If 1 is possible, that seems a little scary (not necessarily for this case, but just in genneral). Is there a way to architecturally guarantee consistency or at least check for it?

2 makes sense. I think we'll eventually move to using events rather than the registry which would make this go away. If we weren't going to do that, we could try to infer the contract type upstream to avoid ingesting incomplete information.

Reply:
> If 1 is possible, that seems a little scary (not just for this case, but just in general), especially when querying from databases that are keyed by time. Is there a way to architecturally guarantee consistency or at least check for it?

I've attempted to do this architecturally in the `api/app.ts` startup sequence. The order in which services get started is important, so we try to run an update to get all the emp state as one of the first routines, and this blocks any other services that would depend on this data from initializing.  Its not fool proof though, as an error fetching state from network can fail, leaving a null field in the emp state, so imo its best to try and be tolerant to inconsistent state.